### PR TITLE
Add pre-commit with various linters and static code checkers

### PR DESCRIPTION
**Pull request by basnijholt**
This adds pre-commit with black, isort, pyupgrade, and a few other nice to haves with a new Github Action that runs pre-commit.

The only changes I made manually are in https://github.com/jupyter/jupyter-sphinx/pull/206/commits/7f35bea0a2123d7e7315a4411992751141751d83 (which also include a few automatically made changes by isort).

### CODE REVIEW

File: jupyter_sphinx/ast.py [hunk 1]
```diff
@@ -45,8 +41,8 @@ def load_content(cell, location, logger):
         try:
             with Path(filename).open() as f:
                 content = [line.rstrip() for line in f.readlines()]
-        except (IOError, OSError):
-            raise IOError("File {} not found or reading it failed".format(filename))
+        except OSError:
+            raise OSError(f"File {filename} not found or reading it failed")
```
Comment (akhmerov) [inline jupyter_sphinx/ast.py:45]: Why the change to the exception types?

Comment (basnijholt) [reply jupyter_sphinx/ast.py:45]: This change is made by pyupgrade and is equivalent because we support Python≥3.6 only.

See https://github.com/asottile/pyupgrade#oserror-aliases

Comment (basnijholt) [reply jupyter_sphinx/ast.py:45]: `isinstance(IOError(), OSError) == True`

**Review by akhmerov — APPROVED** [review]
